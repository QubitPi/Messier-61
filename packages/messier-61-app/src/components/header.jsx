/*
 * Copyright 2023 Paion Data. All rights reserved.
 */
import React from "react";
import { useTranslation } from "react-i18next";

export const Header = (props) => {
  const { t } = useTranslation();

  return (
    <header id="header">
      <div className="intro">
        <div className="overlay">
          <div className="container">
            <div className="row">
              <div className="col-md-8 col-md-offset-2 intro-text">
                <h1>
                  External Brain
                  <span></span>
                </h1>
                <p>{t("Tag Line")}</p>
                <a href="#about" className="btn btn-custom btn-lg page-scroll">
                  Learn More
                </a>{" "}
              </div>
            </div>
          </div>
        </div>
      </div>
    </header>
  );
};
