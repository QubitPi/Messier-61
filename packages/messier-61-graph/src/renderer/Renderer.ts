/*
 * Copyright Jiaqi Liu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { BaseType, Selection } from "d3-selection";
import { Visualization } from "../Visualization";
import { NodeCaptionLine, NodeModel } from "../models/Node";
import { BORDER_COLOR, BORDER_WIDTH, COLOR, FONT_SIZE, TEXT_COLOR_INTERNAL } from "../models/GraphStyle";
import { RelationshipModel } from "../models/Relationship";

type RendererEventHandler<Datum> = (
  selection: Selection<SVGGElement, Datum, BaseType, unknown>,
  style: Visualization
) => any;

const NODE_RING_RENDERER_NAME = "nodeRing";
const NODE_OUTLINE_RENDERER_NAME = "nodeOutline";
const NODE_CAPTION_RENDERER_NAME = "nodeCaption";

const ARROW_PATH_RENDERER_NAME = "arrowPath";
const RELATIONSHIP_TYPE_RENDERER_NAME = "relationshipType";

export const NODE_RING_STROKE_WIDTH_IN_PX = 8;
export const NO_OP_RENDERER_HANDLER = () => undefined;

export class Renderer<Datum> {
  public name: string;
  public onTick: RendererEventHandler<Datum>;
  public onGraphChange: RendererEventHandler<Datum>;

  constructor({name, onTick, onGraphChange}: {name: string, onTick: RendererEventHandler<Datum>, onGraphChange: RendererEventHandler<Datum>}) {
    this.name = name;
    this.onTick = onTick;
    this.onGraphChange = onGraphChange;
  }
}

const NODE_OUTLINE_RENDERER = new Renderer<NodeModel>({
  name: NODE_OUTLINE_RENDERER_NAME,
  onTick: NO_OP_RENDERER_HANDLER,
  onGraphChange(selection, visualizaiton) {
    return selection
      .selectAll("circle.b-outline")
      .data(node => [node])
      .join("circle")
      .classed("b-outline", true)
      .attr("cx", 0)
      .attr("cy", 0)
      .attr("r", (node: NodeModel) => node.radius)
      .attr("fill", (node: NodeModel) => visualizaiton.style.forNode(node).get(COLOR))
      .attr("stroke", (node: NodeModel) => visualizaiton.style.forNode(node).get(BORDER_COLOR))
      .attr("stroke-width", (node: NodeModel) => visualizaiton.style.forNode(node).get(BORDER_WIDTH))
  }
});

const NODE_CAPTION_RENDERER = new Renderer<NodeModel>({
  name: NODE_CAPTION_RENDERER_NAME,
  onTick: NO_OP_RENDERER_HANDLER,
  onGraphChange(selection, visualizaiton) {
    return selection
      .selectAll("text.caption")
      .data((node: NodeModel) => node.caption)
      .join("text")
      // Classed element ensures duplicate data will be removed befoer adding
      .classed("caption", true)
      .attr("text-anchor", 'middle')
      .attr("pointer-events", 'none')
      .attr("x", 0)
      .attr("y", (nodeCaptionLine: NodeCaptionLine) => nodeCaptionLine.baseline)
      .attr(
        "font-size",
        (nodeCaptionLine: NodeCaptionLine) => visualizaiton.style.forNode(nodeCaptionLine.node).get(FONT_SIZE)
      )
      .attr(
        "fill",
        (nodeCaptionLine: NodeCaptionLine) => visualizaiton.style.forNode(nodeCaptionLine.node).get(TEXT_COLOR_INTERNAL)
      )
      .text((nodeCaptionLine: NodeCaptionLine) => nodeCaptionLine.text)
  }
});

/**
 * The hollow ring that shows up around a node when the mouse mouses over a node
 */
const NODE_RING_RENDERER = new Renderer<NodeModel>({
  name: NODE_RING_RENDERER_NAME,
  onTick: NO_OP_RENDERER_HANDLER,
  onGraphChange(selection, visualizaiton) {
    const circles = selection
      .selectAll("circle.ring")
      .data((node: NodeModel) => [node]);

    circles
      .enter()
      .insert("circle", ".b-outline")
      .classed("ring", true)
      .attr("cx", 0)
      .attr("cy", 0)
      .attr("stroke-width", `${NODE_RING_STROKE_WIDTH_IN_PX}px`)
      .attr("r", (node: NodeModel) => node.radius + 4)

    return circles.exit().remove();
  }
});

const ARROW_PATH_RENDERER = new Renderer<RelationshipModel>({
  name: ARROW_PATH_RENDERER_NAME,
  onTick(selection, visualization) {
    return selection
      .selectAll<BaseType, RelationshipModel>("path")
      .attr("d", d => d.arrow!.outline(d.shortCaptionLength ?? 0))
  },
  onGraphChange(selection, visualization) {
    return selection
      .selectAll("path.b-outline")
      .data(relationship => [relationship])
      .join("path")
      .classed("b-outline", true)
      .attr("fill", (relationship: RelationshipModel) => visualization.style.forRelationship(relationship).get(COLOR))
      .attr("stroke", "none")
  }
});

const RELATIONSHIP_TYPE_RENDERER = new Renderer<RelationshipModel>({
  name: RELATIONSHIP_TYPE_RENDERER_NAME,
  onTick(selection, visualizaiton) {
    return selection
      .selectAll<BaseType, RelationshipModel>("text")
      .attr("x", relationship => relationship?.arrow?.midShaftPoint?.x ?? 0)
      .attr("y", relationship => (relationship?.arrow?.midShaftPoint?.y ?? 0) + parseFloat(visualizaiton.style.forRelationship(relationship).get(FONT_SIZE)) / 2 - 1)
      .attr("transform", relationship => {
        if (relationship.naturalAngle < 90 || relationship.naturalAngle > 270) {
          // Calibrate text orientation. For example,
          // Dragging a node from 1st quadrant to the 2nd quadrant will change the text view to upside-down, making the
          // text hard to read. Rotating the text 180 degress around midshaft point will turn it back to the normal-view
          return `rotate(180 ${relationship?.arrow?.midShaftPoint?.x ?? 0} ${relationship?.arrow?.midShaftPoint?.y ?? 0})`
        } else {
          return null;
        }
      })
      .text(relationship => relationship.shortCaption ?? "");
  },
  onGraphChange(selection, visualizaiton) {
    return selection
      .selectAll("text")
      .data(relationship => [relationship])
      .join("text")
      .attr("text-anchor", "middle")
      .attr("pointer-events", "none")
      .attr("font-size", relationship => visualizaiton.style.forRelationship(relationship).get(FONT_SIZE))
      .attr("fill", relationship => visualizaiton.style.forRelationship(relationship).get(`text-color-${relationship.captionLayout}`));
  }
});

export const NODE_RENDERERS = [NODE_OUTLINE_RENDERER, NODE_CAPTION_RENDERER, NODE_RING_RENDERER];
