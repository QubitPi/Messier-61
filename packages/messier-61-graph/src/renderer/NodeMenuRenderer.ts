import { BaseType, Selection } from "d3-selection";
import { Arc, DefaultArcObject, arc as d3Arc } from "d3-shape";
import { NODE_DOUBLE_CLICKED } from "../event-handler/GraphEventHandlerModel";
import { NodeModel } from "../models/Node";
import { NODE_RING_STROKE_WIDTH_IN_PX, NO_OP_RENDERER_HANDLER, Renderer } from "./Renderer";
import { Visualization } from "../Visualization";
import { CONTEXT_MENU_ITEM } from "../VizItem";

/*
 * Copyright Jiaqi Liu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const DONUT_REMOVE_NODE_RENDERER_NAME = "donutRemoveNode";
const DONUT_EXPAND_NODE_RENDERER_NAME = "donutExpandNode";
const DONUT_UNLOCK_NODE_RENDERER_NAME = "donutUnlockNode";

const ITEM_INDEX_MAP = new Map<string, number>([
  [DONUT_REMOVE_NODE_RENDERER_NAME, 1],
  [DONUT_EXPAND_NODE_RENDERER_NAME, 2],
  [DONUT_UNLOCK_NODE_RENDERER_NAME, 3],
]);

const MENU_ANGLE = (2 * Math.PI) / ITEM_INDEX_MAP.size;

const donutExpandNode = new Renderer<NodeModel>({
  name: DONUT_EXPAND_NODE_RENDERER_NAME,
  onTick: NO_OP_RENDERER_HANDLER,
  onGraphChange(selection, visualizaiton) {
    return createMenuItem(selection, visualizaiton, NODE_DOUBLE_CLICKED, 2, NODE_EXPAND_CLASS_NAME, [-8, -10]);
  },
});

export const EXPAND_COLLAPSE_ICON_KEY = "Expand / Collapse";
export const UNLOCK_ICON_KEY = "Unlock";
export const REMOVE_ICON_KEY = "Remove";

export const NODE_EXPAND_CLASS_NAME = "expand-node";

/**
 * Attatch node menues, 
 * @param selection 
 * @param visualizaiton 
 * @param eventType 
 * @param item 
 * @param className 
 * @param position 
 * @param svgIconKey 
 * @param tooltip 
 */
function createMenuItem(
  selection: Selection<SVGGElement, NodeModel, BaseType, unknown>,
  visualizaiton: Visualization,
  eventType: string,
  item:
    | typeof DONUT_REMOVE_NODE_RENDERER_NAME
    | typeof DONUT_EXPAND_NODE_RENDERER_NAME
    | typeof DONUT_UNLOCK_NODE_RENDERER_NAME,
  className: string,
  position: [number, number],
  svgIconKey: typeof EXPAND_COLLAPSE_ICON_KEY | typeof UNLOCK_ICON_KEY | typeof REMOVE_ICON_KEY,
  tooltip: string
) {
  const tab = selection
    .selectAll(`path.${className}`)
    .data(getSelectedNode)
    .join("path")
    .classed(className, true)
    .classed(CONTEXT_MENU_ITEM, true)
    .attr("d", (node) => drawArc(node.radius, item)());

  const rawSvgIcon = 
}

function getSelectedNode(node: NodeModel) {
  return node.selected ? [node] : [];
}

const DEFAULT_WIDTH = 30;

/**
 * Returns a **function** that draws a menu option around a node using [D3 arc](https://github.com/d3/d3-shape#arcs).
 * 
 * When the return value of this function is used as `someFunction` in `.attr("d", someFunction)`, it is very important
 * to remember that the 2nd argument of `.attr()` must either be
 * [constants or functions](https://www.d3indepth.com/selections/#updating-selections-with-functions), therefore we
 * must end the call with an extra parenthesis, i.e.
 * 
 * ```typescript
 * .attr("d", (node) => drawArc(node.radius, item)());
 * ```
 * 
 * Note the last `()` after the `drawArc(node.radius, item)`, We are calling the function returned by this method.
 * i.e. `fn()`, where `fn = drawArc(node.radius, item)`
 * 
 * @param nodeRadius The radius of the enclusing node
 * @param item The menu type/name to draw arc for. Can only be one of the
 * {@link !DONUT_REMOVE_NODE_RENDERER_NAME}/{@link !DONUT_EXPAND_NODE_RENDERER_NAME}/{@link !DONUT_UNLOCK_NODE_RENDERER_NAME}
 * @param arcWidth The difference between outer arc radius and inner ard radius. Defaults to {@link !DEFAULT_WIDTH}.
 *
 * @returns a [D3 arc](https://github.com/d3/d3-shape#arcs) instance
 */
function drawArc(
  nodeRadius: number,
  item:
    | typeof DONUT_REMOVE_NODE_RENDERER_NAME
    | typeof DONUT_EXPAND_NODE_RENDERER_NAME
    | typeof DONUT_UNLOCK_NODE_RENDERER_NAME,
  arcWidth = DEFAULT_WIDTH
): any {
  const startAngle = MENU_ANGLE * ITEM_INDEX_MAP.get(item)!;
  const endAngle = startAngle + MENU_ANGLE;

  const innerRadius = Math.max(nodeRadius + NODE_RING_STROKE_WIDTH_IN_PX, 20);

  return d3Arc()
    .innerRadius(innerRadius)
    .outerRadius(innerRadius + arcWidth)
    .startAngle(startAngle)
    .endAngle(endAngle)
    .padAngle(0.03)!; // https://github.com/d3/d3-shape#arc_padAngle
}
