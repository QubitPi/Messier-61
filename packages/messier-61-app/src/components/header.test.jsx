/*
 * Copyright 2023 Paion Data. All rights reserved.
 */
import renderer from "react-test-renderer";
import { Header } from "./header";

it("header part renders correctly", () => {
  const tree = renderer.create(<Header />).toJSON();
  expect(tree).toMatchSnapshot();
});
