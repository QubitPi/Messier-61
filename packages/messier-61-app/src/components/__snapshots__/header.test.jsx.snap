// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header part renders correctly 1`] = `
<header
  id="header"
>
  <div
    className="intro"
  >
    <div
      className="overlay"
    >
      <div
        className="container"
      >
        <div
          className="row"
        >
          <div
            className="col-md-8 col-md-offset-2 intro-text"
          >
            <h1>
              External Brain
              <span />
            </h1>
            <p>
              A completely new way for personal knowledge management for young ages
            </p>
            <a
              className="btn btn-custom btn-lg page-scroll"
              href="#about"
            >
              Learn More
            </a>
             
          </div>
        </div>
      </div>
    </div>
  </div>
</header>
`;
